### PR TITLE
Tidy up code

### PR DESCRIPTION
**Pull request by EdAbati**
This PR tries to tackle #11. objective: tidy up code and add more documentation towards  (please bear in mind some stuff may still be missing for example documentation and mention of W&B).

- removed everything (variables/ functions) that is not used (if something is needed should be added in a next PR, together with the code that uses it )
- changed structure to resurface main code (cli, model training, notebooks) that may be a bit lost in the folders. `notebooks` and `app` (and everything that is has to do with it) are in root now. the Jupyter service still uses their content as before.
- simplified (hopefully) docker files and requirements removing what it is not needed (I didn't know how to add stuff to `poetry` I went with what was taught in the course `Conda`+`pip-tools`)
- collated documentations and readmes
- added documentation on how to update services individually (e.g. just add stuff to `streamlit`)
- fixed few issues with UI and model training 

Things to finish/discuss:
- save torch script model somewhere that is not local. I removed S3 because it was not documented and it adds yet another tool/dependency. I personally feel that this is not a strict requirement, but if we want to add the artifact somewhere wee should use W&B for this too (since we use it already)
- restore `poetry`? Let me know if you preferred `poetry` and I will go back to it, I just need to find out out to simply add or remove requirements. :)

### CODE REVIEW

**Comment by diegoquintanav** [comment]
There’s a lot going on here.

## About the folder structure

1. I see that you moved the contents of the `jupyter` container to the root level. I think this makes the project look messier. Why do you think this is needed? I think leaving it in the `services` folder makes things more clear, otherwise, it is just moving files around in a setting that was already working. I don't think this is productive.

## About `conda, pip, and poetry`

1. I see you replace `poetry` for `conda` in the `jupyter` container. `conda` depends on other repositories like `conda-forge` to install dependencies. When a dependency is not available in `conda` one must resort to pip and that breaks the dependency management of `conda` entirely. Poetry on the other hand, uses pip only. But more important than that, `Poetry`  was already set up and working. Why change it? 
1. In addition to the previous point, `conda` development flow inside containers requires prepending `conda run -n environment` to run things inside the environment. Otherwise, you need to  `activate` the environment, which requires human interaction, which is counterintuitive in a container setting. In the already implemented containers, the `PYTHONPATH` was already set and working without any kind of activation step. 
1. I see some calls to pip-tools and pip-sync. What are these doing? Poetry is a container dependency, now I have to install two tools in the docker host?

## About S3

1. The S3 step was because it detaches the state of your local repository from the container. I prefer a CLI command inside the `jupyter` services that trains a model, compares it against the existing one, and pushes the new version to a remote location so now it’s available to everyone. You already do this with your CLI except for the uploading part, meaning that I need access to your filesystem to build a new image (it had a `COPY model.pt .`). 
2. Moving this CLI to the jupyter container makes triggering the flow from the container and not from the docker host, and it was this what I meant during a conversation we had.
3. I agree we could use `wandb` to fetch the model and trace it in the `base` step of the docker image (it is currently a multistage build with two steps and one dependency). To do so you need now `torch` and `wandb` dependencies instead of just `boto3`. Why do this if we can (and did) already upload the model to `s3` in the `jupyter` service?

## About traefik

1. I removed traefik it in another PR because it was not fully implemented. In a production setting though, a reverse proxy comes very handy, because opening ports inside machines may not be an option. Forwarding ports using `vscode` works only if you are the only one using the forwarded port. Others won’t be able to access your `localhost` if they are not on the same network as you.

**Comment by EdAbati** [comment]
TLDR: I can see that you are not comfortable with the changes. Since we don’t have time for discussion, I will close this and reopen another one with just the fixes to the code that doesn’t work. But _please document_ the decisions you took and remove what you deprecated then.  

## About folder structure

> I see that you moved the contents of the jupyter container to the root level. I think this makes the project look messier. 

Well there are currently 2 locations: there is some code in root and some code inside `services`.  I just tried to collate it all in 1 place (IMO one would expect to see the training CLI code in root instead of inside a service folder called `Jupiter`).  But no worries we can do it the way that you prefer, there is no much time to change this.  

## About Conda, pip and poetry

> But more important than that, Poetry was already set up and working. Why change it?

We had conda with pip working, until your last PR. Why did you change it? :)
The use of `conda` and `pip-tools` is the method suggested in the course and I thought everyone would have been familiar with it.
If now we go with poetry as you want, please document it because there is no mention of poetry in any readme and there are still references of Conda (e.g. in the makefile and in `Conda-environment.yml`).  

## About S3

> I prefer a CLI command inside the jupyter services that trains a model, compares it against the existing one, and pushes the new version to a remote location so now it’s available to everyone.

We don't have that. Is the remote location available to everyone in the team?
 If we had something that compares the models automatically, I could understand your point.  At the moment, we just have models artifacts in Wandb.  Currently one has to pick the favourite model and trigger the CLI manually to register it and convert it to torch script. Therefore, at this point, one can just ship it directly.  
 Anyway, if you want to add the extra step of uploading and downloading from/to S3 please add a readme, that explains your decision.
 
##  About traefik  
 
 I didn’t touch it.

**Comment by diegoquintanav** [comment]
> TLDR: I can see that you are not comfortable with the changes. Since we don’t have time for discussion, I will close this and reopen another one with just the fixes to the code that doesn’t work. But please document the decisions you took and remove what you deprecated then.
> 

I documented things as much as I could, and I asked you about where I could improve this, and I think I did. Or not. If you had troubles with poetry (e.g. not knowing how to add a package) you could have asked me or the poetry documentation before changing everything again. I agree I didn't ask before moving to poetry before, and here I explained my reasoning.

> Well there are currently 2 locations: there is some code in root and some code inside services. I just tried to collate it all in 1 place (IMO one would expect to see the training CLI code in root instead of inside a service folder called Jupiter).
>

This redundancy is because we were working in different ways. In another PR I suggested moving this CLI eventually to the jupyter service. At the moment this has zero relevance to me because as you say, we don't have much time. It's a matter of taste and we are allowed to differ on that. The thing is, If it helps you get things done I won't touch it. I don’t need it. We can change it later. Perhaps a bit of chat before changing everything I was using would have helped tons here.

> But no worries we can do it the way that you prefer, there is no much time to change this.
>

I find this is uncalled for. This PR is actually doing that: it changes the code the way *you* prefer without any kind of real discussion, removing stuff I am using, at a point where we have other priorities. This has already used *your* time to do all these changes, and *my* time to address them.

> We had conda with pip working, until your last PR. Why did you change it? :)
The use of conda and pip-tools is the method suggested in the course and I thought everyone would have been familiar with it.
If now we go with poetry as you want, please document it because there is no mention of poetry in any readme and there are still references of Conda (e.g. in the makefile and in Conda-environment.yml).
> 

See my answer to the first paragraph. I don't think replicating the `poetry` documentation here is very productive. Those traces of conda are because of the 2 locations we had before (see my answer to your second paragraph). I don't want to break what was working for you (or others, I really don't know) and I think this could be addressed with some sort of discussion, with a PR that settles this.

> We don't have that. Is the remote location available to everyone in the team?
If we had something that compares the models automatically, I could understand your point. At the moment, we just have models artifacts in Wandb. Currently one has to pick the favourite model and trigger the CLI manually to register it and convert it to torch script. Therefore, at this point, one can just ship it directly.
Anyway, if you want to add the extra step of uploading and downloading from/to S3 please add a readme, that explains your decision.
> 

Yes. the location is available to everyone on the team, I left my AWS credentials in the lambdalabs machine and [mentioned it in the discord chat](https://discordapp.com/channels/984525101678612540/1015151891564474379/1028713763135819887).  This code was added in another PR and we could have discussed it there. If you merged it I assume you are ok with it. Again, this could be solved by asking. 

That model deployment workflow is something we don’t have but I want to work towards it, and leaving some unused methods may not agree with your view of "tidy" code but hey, can you chill and ignore it, perhaps not delete it? This repo is in a very early stage and all of this just adds friction. 

> I didn't touch it
> 

[I have to disagree with this](https://github.com/EdAbati/fsdl-2022-weak-supervision-project/pull/18/commits/38e2f6f727b6783e35f611cff16e216daf4e94a3). I removed it on another PR that is already merged so it does not matter, really.

## Final note

I'm ok to merge all of this if it helps the team, of course. or the other PR. All that I am saying is: This is a big PR that removes stuff I use and it would have helped a bit of discussion before, or perhaps splitting it into smaller PRs?

**Comment by EdAbati** [comment]
Sorry I may have sounded more harsh than I meant and apologies for that.

Since we didn't have time for discussing, the PR was just a way to 'ask' not a way to 'force the decision in any way'. (otherwise I would have just silently pushed it on main 😅) That's why I was happy to close it, really.

My point was since a lot of my workflow broke with the recent big PR I wanted to find a way to merge the 2 way of working so everyone was happy.
And sorry also for deleting your unused code, but there was not much explanation that 'you wanted to work towards it'. In my view it was just forgotten.

Hope there are no hard feelings.
